File: src/main/java/work/lclpnet/ap2/api/actor/ActorData.java
package work.lclpnet.ap2.api.actor;

import com.mojang.serialization.Codec;
import com.mojang.serialization.DynamicOps;

public record ActorData<D>(D data, Codec<D> codec) {

    public <T> void encode(DynamicOps<T> ops, T prefix) {
        codec.encode(data, ops, prefix);
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/api/actor/ActorData.kt
package work.lclpnet.ap2.api.actor

import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import com.mojang.serialization.DynamicOps

@JvmRecord
data class ActorData<D>(val data: D, val codec: Codec<D>) {
    fun <T> encode(ops: DynamicOps<T>, prefix: T): DataResult<T> =
        codec.encode<T>(data, ops, prefix)
}
